warp/Network/Wai/Handler/Warp/File.hs

```diff
@@ -0,0 +1,173 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE BangPatterns #-}
+
+module Network.Wai.Handler.Warp.File (
+    RspFileInfo(..)
+  , conditionalRequest
+  , addContentHeadersForFilePart
+  ) where
+
+import Control.Applicative ((<|>))
+import Data.Array ((!))
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as B hiding (pack)
+import qualified Data.ByteString.Builder as B
+import qualified Data.ByteString.Char8 as B (pack, readInteger)
+import qualified Data.ByteString.Lazy as L
+import Data.Maybe
+import Network.HTTP.Date
+import qualified Network.HTTP.Types as H
+import qualified Network.HTTP.Types.Header as H
+import Network.Wai
+import qualified Network.Wai.Handler.Warp.FileInfoCache as I
+import Network.Wai.Handler.Warp.Header
+import Numeric (showInt)
+
+----------------------------------------------------------------
+
+data RspFileInfo = WithoutBody H.Status
+                 | WithBody H.Status H.ResponseHeaders Integer Integer
+                 deriving (Eq,Show)
+
+----------------------------------------------------------------
+
+conditionalRequest :: I.FileInfo
+                   -> H.ResponseHeaders -> IndexedHeader
+                   -> RspFileInfo
+conditionalRequest finfo hs0 reqidx = case fromJust mx of
```

Comment:
I'd rather not include a `fromJust` here. My best preference is to make the call total by making `uncoditional` a fallback (since it appears that it always returns a `Just` value). My second-best preference would be to pattern match on the `Maybe` and use an explicit `error` call that states that `conditionalRequest` failed, since getting a `fromJust` error from a program is maddening.